### PR TITLE
add 10b experiment to flava native and fix checkpoint wrapper

### PR DESCRIPTION
**Pull request by edward-io**
Summary:

Adds 10B experiment config to FLAVA native training script and fixes issues with checkpointing due to kwargs and re-entrant.

Test plan:

`torchrun --nproc_per_node=8 -m flava.native.train config=flava/native/configs/10b.yaml`

Fixes #{issue number}

### CODE REVIEW

**Comment by langong347** [comment]
I know this is still a draft version. Would you mind putting a proper summary and a test plan so members from the team who do not have prior context about this work-stream can understand. Thanks :)

File: examples/flava/native/train.py [hunk 1]
```diff
@@ -167,34 +193,16 @@ def create_model(self) -> torch.nn.Module:
                 mixed_precision=mp,
                 auto_wrap_policy=partial(
                     transformer_auto_wrap_policy,
-                    transformer_layer_cls={TransformerEncoderLayer},
```
Comment (ankitade) [inline examples/flava/native/train.py:170]: can we add a toggle for this?

File: examples/flava/native/train.py [hunk 2]
```diff
@@ -143,6 +146,29 @@ def create_model(self) -> torch.nn.Module:
 
         model = model.to(self.device)
         print0(f"after moving to cuda: {torch.cuda.memory_allocated()/1024**3:.3} GB")
+
+        if self.config.training.activation_checkpointing:
+            # note: activation checkpointing wrapper currently is faulty
+            # - non-reentrant does not support kwargs in TransformerEncoderLayer
+            # - memory reduction from checkpointing is less than expected
```
Comment (ankitade) [inline examples/flava/native/train.py]: can remove comments now?

File: examples/flava/native/train.py [hunk 3]
```diff
@@ -143,6 +146,25 @@ def create_model(self) -> torch.nn.Module:
...
+                checkpoint_impl = CheckpointImpl.REENTRANT
+            else:
+                checkpoint_impl = CheckpointImpl.NO_REENTRANT
+
+            non_reentrant_wrapper = partial(
+                checkpoint_wrapper,
+                offload_to_cpu=False,
+                checkpoint_impl=checkpoint_impl,
+            )
+            apply_activation_checkpointing_wrapper(
```
Comment (rohan-varma) [inline examples/flava/native/train.py:162]: I don't like that this is the state that we're in, but can we but apply_ac_wrapper after FSDP wrap if using non-reentrant, and before FSDP wrap using reentrant? This is the way that both of them would work (the difference is not good).

File: examples/flava/native/configs/10b.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
+training:
```
Comment (ankitade) [inline examples/flava/native/configs/10b.yaml:1]: do u want to check in the 4.8b one while you are at it

File: examples/flava/native/train.py [hunk 5]
```diff
@@ -133,7 +136,7 @@ def create_model(self) -> torch.nn.Module:
         model_config = self.config.get("model", {})
         print0(f"using model config: {model_config}")
 
-        model = FLAVAPreTrainModule(**model_config)
+        model = FLAVAPreTrainModule(**model_config).cuda()
```
[the remark this answers is not in the document]
Comment (edward-io) [reply examples/flava/native/train.py]: i removed this line, i don't think it speeds up the model construction

File: examples/flava/native/configs/10b.yaml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  strategy: fsdp # can be changed to ddp or fsdp
+  seed: 1337
+
+  batch_size: 16
```
Comment (ankitade) [inline examples/flava/native/configs/10b.yaml:5]: just make it 8 for consistency

File: examples/flava/native/train.py (hunk 1 above)
Comment (edward-io) [reply examples/flava/native/train.py:170]: Sure, what are the two choices? It needs some wrapping policy AFAIK.

File: examples/flava/native/train.py (hunk 3 above)
Comment (edward-io) [reply examples/flava/native/train.py:162]: do we want to keep non-reentrant if the memory usage is greater than reentrant?

File: examples/flava/native/configs/10b.yaml (hunk 4 above)
Comment (edward-io) [reply examples/flava/native/configs/10b.yaml:1]: Sure!

File: examples/flava/native/train.py (hunk 3 above)
Comment (rohan-varma) [reply examples/flava/native/train.py:158]: Good point, let's just remove non-reentrant.

File: examples/flava/native/train.py [hunk 7]
```diff
@@ -165,36 +189,17 @@ def create_model(self) -> torch.nn.Module:
             model = FSDP(
                 model,
                 mixed_precision=mp,
+                device_id=self.device,
                 auto_wrap_policy=partial(
                     transformer_auto_wrap_policy,
-                    transformer_layer_cls={TransformerEncoderLayer},
+                    transformer_layer_cls={
+                        TransformerEncoderLayer,
```
Comment (ankitade) [inline examples/flava/native/train.py:196]: remove

Comment (edward-io) [reply examples/flava/native/train.py:196]: @ankitade  i thought we wanted to keep this for the imagenet validation?

File: examples/flava/native/configs/4.8b.yaml [hunk 8]
```diff
@@ -0,0 +1,79 @@
+training:
+  strategy: fsdp # can be changed to ddp or fsdp
+  seed: 1337
+
+  batch_size: 12
```
Comment (ankitade) [inline examples/flava/native/configs/4.8b.yaml:5]:
```suggestion
  batch_size: 8
```

**Review by ankitade — APPROVED** [review]

**Pull request closed without merging**